src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs

```diff
@@ -52,465 +54,314 @@ public class AllSubscription : ISubscriptionEnumerator {
+								return;
+							}
 
-					if (_disposedTokenSource.IsCancellationRequested) {
-						return false;
+							ReadPage(Position.FromInt64(
+								completed.NextPos.CommitPosition,
+								completed.NextPos.PreparePosition), OnMessage);
+							return;
+						case ReadAllResult.AccessDenied:
+							_channel.Writer.TryComplete(RpcExceptions.AccessDenied());
+							return;
+						default:
+							_channel.Writer.TryComplete(RpcExceptions.UnknownError(completed.Result));
+							return;
 					}
+				}
+			}
 
-					goto ReadLoop;
+			private void GoLive(Position startPosition) {
+				var liveEvents = Channel.CreateBounded<ResolvedEvent>(BoundedChannelOptions);
+				var caughtUpSource = new TaskCompletionSource<TFPos>();
+				var liveMessagesCancelled = 0;
 
-					void OnMessage(Message message) {
-						if (message is ClientMessage.NotHandled notHandled &&
-						    RpcExceptions.TryHandleNotHandled(notHandled, out var ex)) {
-							readNextSource.TrySetException(ex);
-							return;
-						}
+				Log.Information(
+					"Live subscription {subscriptionId} to $all running from {position}...",
+					_subscriptionId, startPosition);
 
-						if (!(message is ClientMessage.ReadAllEventsForwardCompleted completed)) {
-							readNextSource.TrySetException(
-								RpcExceptions.UnknownMessage<ClientMessage.ReadAllEventsForwardCompleted>(message));
-							return;
-						}
+				_bus.Publish(new ClientMessage.SubscribeToStream(Guid.NewGuid(), _subscriptionId,
+					new ContinuationEnvelope(OnSubscriptionMessage, _semaphore, _cancellationToken), _subscriptionId,
+					string.Empty, _resolveLinks, _user));
 
-						switch (completed.Result) {
-							case ReadAllResult.Success:
-								foreach (var @event in completed.Events) {
-									var position = Position.FromInt64(
-										@event.OriginalPosition.Value.CommitPosition,
-										@event.OriginalPosition.Value.PreparePosition);
-									if (position <= _startPosition) {
-										Log.Verbose(
-											"Catch-up subscription {subscriptionId} to $all skipped event {position}.",
-											_subscriptionId, position);
-										continue;
-									}
-
-									Log.Verbose(
-										"Catch-up subscription {subscriptionId} to $all received event {position}.",
-										_subscriptionId, position);
-
-									_buffer.Enqueue(@event);
-								}
+				Task.Factory.StartNew(PumpLiveMessages, _cancellationToken);
 
-								_nextPosition = Position.FromInt64(
-									completed.NextPos.CommitPosition,
-									completed.NextPos.PreparePosition);
-								readNextSource.TrySetResult(completed.IsEndOfStream);
-								return;
-							case ReadAllResult.AccessDenied:
-								readNextSource.TrySetException(RpcExceptions.AccessDenied());
-								return;
-							default:
-								readNextSource.TrySetException(RpcExceptions.UnknownError(completed.Result));
-								return;
-						}
+				async Task PumpLiveMessages() {
+					await caughtUpSource.Task.ConfigureAwait(false);
```

Comment:
It seems there are some cases where this task will wait forever (i will refer to this comment below)

Reply:
I've not added all the cases - one way to fix would be to call caughtUpSource.TrySetException and handle it here.

Reply:
or maybe a simpler way: get rid of the caughtUpSource task and start the PumpLiveMessages task only after NotifyCaughtUp() is called.

Reply:
Did it via `caughtUpSource.TrySetException`